l2-contracts/contracts/bridge/L2NativeTokenVault.sol

```diff
@@ -0,0 +1,187 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.20;
+
+import {Ownable2StepUpgradeable} from "@openzeppelin/contracts-upgradeable/access/Ownable2StepUpgradeable.sol";
+import {BeaconProxy} from "@openzeppelin/contracts/proxy/beacon/BeaconProxy.sol";
+import {UpgradeableBeacon} from "@openzeppelin/contracts/proxy/beacon/UpgradeableBeacon.sol";
+
+import {IL2SharedBridge} from "./interfaces/IL2SharedBridge.sol";
+import {IL2StandardToken} from "./interfaces/IL2StandardToken.sol";
+import {IL2NativeTokenVault} from "./interfaces/IL2NativeTokenVault.sol";
+
+import {L2StandardERC20} from "./L2StandardERC20.sol";
+import {L2ContractHelper, DEPLOYER_SYSTEM_CONTRACT, NATIVE_TOKEN_VAULT_VIRTUAL_ADDRESS, IContractDeployer} from "../L2ContractHelper.sol";
+import {SystemContractsCaller} from "../SystemContractsCaller.sol";
+
+import {EmptyAddress, EmptyBytes32, AddressMismatch, Bytes32Mismatch, DeployFailed, AmountMustBeGreaterThanZero, InvalidCaller} from "../L2ContractErrors.sol";
+
+/// @author Matter Labs
+/// @custom:security-contact security@matterlabs.dev
+/// @notice The "default" bridge implementation for the ERC20 tokens. Note, that it does not
+/// support any custom token logic, i.e. rebase tokens' functionality is not supported.
+contract L2NativeTokenVault is IL2NativeTokenVault, Ownable2StepUpgradeable {
+    IL2SharedBridge public override l2Bridge;
+
+    /// @dev Contract that stores the implementation address for token.
+    /// @dev For more details see https://docs.openzeppelin.com/contracts/3.x/api/proxy#UpgradeableBeacon.
+    UpgradeableBeacon public l2TokenBeacon;
+
+    /// @dev Bytecode hash of the proxy for tokens deployed by the bridge.
+    bytes32 internal l2TokenProxyBytecodeHash;
+
+    mapping(bytes32 assetId => address tokenAddress) public override tokenAddress;
+
+    modifier onlyBridge() {
+        if (msg.sender != address(l2Bridge)) {
+            revert InvalidCaller(msg.sender);
+            // Only L2 bridge can call this method
+        }
+        _;
+    }
+
+    /// @dev Contract is expected to be used as proxy implementation.
+    /// @dev Disable the initialization to prevent Parity hack.
+    constructor() {
+        _disableInitializers();
+    }
+
+    /// @dev Sets the L1ERC20Bridge contract address. Should be called only once.
+    function setSharedBridge(IL2SharedBridge _sharedBridge) external onlyOwner {
```

Comment:
Also, maybe put it inside the initializer? To not have to deal with multiple L1->L2 txs

Reply:
It is alread inside the L2SharedBridge initializer. It will be upgraded on mainnet, but locally that is the call that deploys the contracts.